File: C++/construct-the-minimum-bitwise-array-i.cpp
// Time:  O(n)
// Space: O(1)

// bit manipulation
class Solution {
public:
    vector<int> minBitwiseArray(vector<int>& nums) {
        vector<int> result(size(nums), -1);
        for (int i = 0; i < size(nums); ++i) {
            if (nums[i] & 1) {
                result[i] = nums[i] - (((nums[i] + 1) & (~nums[i])) >> 1);
            }
        }
        return result;
    }
};

// Time:  O(n * r)
// Space: O(1)
// brute force
class Solution2 {
public:
    vector<int> minBitwiseArray(vector<int>& nums) {
        vector<int> result(size(nums), -1);
        for (int i = 0; i < size(nums); ++i) {
            for (int j = 0; j < nums[i]; ++j) {
                if ((j | (j + 1)) == nums[i]) {
                    result[i] = j;
                    break;
                }
            }
        }
        return result;
    }
};

File: Python/construct-the-minimum-bitwise-array-i.py
# Time:  O(n)
# Space: O(1)

# bit manipulation
class Solution(object):
    def minBitwiseArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        return [x-(((x+1)&~x)>>1) if x&1 else -1 for x in nums]


# Time:  O(n * r)
# Space: O(1)
# brute force
class Solution2(object):
    def minBitwiseArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        return [next((i for i in xrange(x) if i|(i+1) == x), -1) for x in nums]
